### PR TITLE
fix layout transform

### PR DESCRIPTION
**Pull request by dzhwinter**

### CODE REVIEW

File: paddle/framework/data_transform.cc [hunk 1]
```diff
@@ -64,7 +65,7 @@ void TransDataType(const platform::DeviceContext* ctx,
       framework::VisitDataType(dst_type, CastDataType<int>(src, dst, ctx));
       break;
     case proto::DataType::INT64:
-      framework::VisitDataType(dst_type, CastDataType<int64_t>(src, dst, ctx));
+      framework::VisitDataType(dst_type, CastDataType<int>(src, dst, ctx));
```
Comment (chengduoZH) [inline paddle/framework/data_transform.cc]: why do you change `int64_t ` to `int`?

File: paddle/framework/data_transform.cc [hunk 2]
```diff
@@ -76,12 +77,12 @@ void TransDataType(const platform::DeviceContext* ctx,
 
 void TransDataLayout(const platform::DeviceContext* ctx,
                      const KernelTypePair& kernel_pair, const Variable& in,
-                     Variable* out) {
+                     Variable* out, const std::vector<int>& axis) {
```
Comment (chengduoZH) [inline paddle/framework/data_transform.cc]: Parameter order should be: inputs, then outputs.

File: paddle/framework/data_transform.cc [hunk 3]
```diff
@@ -91,9 +92,8 @@ void TransDataLayout(const platform::DeviceContext* ctx,
   dst->Resize(src_dim);
   auto place = kernel_pair.second.place_;
   CopyFrom(src, place, *ctx, dst);
-  const std::vector<int> axis = {0, 2, 3, 1};
 
-  std::vector<int64_t> dst_dim;
+  std::vector<int> dst_dim;
```
Comment (chengduoZH) [inline paddle/framework/data_transform.cc]: The same question: `int64_t`->`int`.

File: paddle/framework/data_transform_test.cc [hunk 4]
```diff
@@ -127,7 +127,9 @@ TEST(DataTransform, Layout) {
   }
 
   Tensor dst = out.Get<Tensor>();
-  EXPECT_TRUE(dst.layout() != src->layout());
+
+  EXPECT_TRUE(dst.layout() == DataLayout::kNCHW);
+  EXPECT_TRUE(dst.dims() == make_ddim({2, 2, 3, 1}));
```
Comment (chengduoZH) [inline paddle/framework/data_transform_test.cc]: The unit test of `NCHW -> NHWC` is also need.

File: paddle/framework/data_transform.cc [hunk 5]
```diff
@@ -111,5 +111,18 @@ void TransDataLayout(const platform::DeviceContext* ctx,
 }  // namespace paddle
 
 namespace f = paddle::framework;
+
+std::vector<int> NHWC2NCHW = {0, 3, 1, 2};
+std::vector<int> NCHW2NHWC = {0, 2, 3, 1};
```
Comment (chengduoZH) [inline paddle/framework/data_transform.cc]: Is it more appropriate to put this in an anonymous namespace?

File: paddle/framework/data_transform.cc (hunk 2 above)
Comment (dzhwinter) [reply paddle/framework/data_transform.cc]: done

File: paddle/framework/data_transform.cc (hunk 3 above)
Comment (dzhwinter) [reply paddle/framework/data_transform.cc]: a mistake change, done.

File: paddle/framework/data_transform_test.cc (hunk 4 above)
Comment (dzhwinter) [reply paddle/framework/data_transform_test.cc]: done.

File: paddle/framework/data_transform.cc (hunk 5 above)
Comment (dzhwinter) [reply paddle/framework/data_transform.cc]: .cc file always safe with translate unit local object. But you are right, anonymous namespace or static keywords is better. 
Done.

File: paddle/framework/data_transform.cc [hunk 6]
```diff
@@ -91,7 +93,6 @@ void TransDataLayout(const platform::DeviceContext* ctx,
   dst->Resize(src_dim);
   auto place = kernel_pair.second.place_;
   CopyFrom(src, place, *ctx, dst);
```
Comment (chengduoZH) [inline paddle/framework/data_transform.cc]: `CopyFrom(src, place, *ctx, dst);` is unnecessary.

File: paddle/framework/data_transform.cc [hunk 7]
```diff
@@ -111,5 +112,22 @@ void TransDataLayout(const platform::DeviceContext* ctx,
 }  // namespace paddle
```
Comment (chengduoZH) [inline paddle/framework/data_transform.cc]: Parameter order should be inputs then outputs: line 106 `CastDataLayout(src, dst, ctx, axis)` -> `CastDataLayout( ctx, axis,src, dst)`

File: paddle/framework/data_transform.cc [hunk 8]
```diff
@@ -74,14 +75,15 @@ void TransDataType(const platform::DeviceContext* ctx,
...
+void TransDataLayout(const std::vector<int>& axis,
+                     const platform::DeviceContext* ctx,
                      const KernelTypePair& kernel_pair, const Variable& in,
                      Variable* out) {
   PADDLE_ENFORCE(in.IsType<Tensor>(), "Only Support Tensor transform!.");
   PADDLE_ENFORCE(
       platform::places_are_same_class(kernel_pair.first.place_,
                                       kernel_pair.second.place_),
-      "TransDataType Only Support DataType transform on same place!");
+      "TransDataLayout Only Support DataType transform on same place!");
```
[the remark this answers is not in the document]
Comment (dzhwinter) [reply paddle/framework/data_transform.cc]: done.

File: paddle/framework/data_transform.cc (hunk 7 above)
Comment (dzhwinter) [reply paddle/framework/data_transform.cc]: done.

File: paddle/framework/data_transform_test.cc [hunk 9]
```diff
@@ -127,7 +127,19 @@ TEST(DataTransform, Layout) {
   }
 
   Tensor dst = out.Get<Tensor>();
```
Comment (chengduoZH) [inline paddle/framework/data_transform_test.cc]: When the Variable is on the GPU, it should also be tested in this unit test.

File: paddle/framework/data_transform.cc [hunk 10]
```diff
@@ -74,26 +75,28 @@ void TransDataType(const platform::DeviceContext* ctx,
...
+                 "TransDataLayout only support Datatype are same!");
 
   auto src = in.Get<Tensor>();
   auto* dst = out->GetMutable<Tensor>();
   PADDLE_ENFORCE(arity(src.dims()) == 4, "Input Arity Only Suppport 4!");
 
-  auto src_dim = src.dims();
-  dst->Resize(src_dim);
   auto place = kernel_pair.second.place_;
   CopyFrom(src, place, *ctx, dst);
```
Comment (chengduoZH) [inline paddle/framework/data_transform.cc]: `CopyFrom` should be replaced with `dst->mutable_data(place, src.type());`

**Comment by chengduoZH** [comment]
Please fix follow those comments in next PR.

File: paddle/framework/data_transform_test.cc (hunk 9 above)
Comment (dzhwinter) [reply paddle/framework/data_transform_test.cc]: Thanks for pointing these flaws out! These fix will be done in next PR.

**Pull request merged**